Adding Content to Statements Chapter...

mattrips opened the pull request:
... added styling, auto numbering, etc.

First and foremost, I'm offering up some content in the Statements chapter.  Checkout Condition Lists and If Statements.

Please review the styles added, and in particular, the indentation and the auto numbering of sections, subsections, sub subsections, and information items.

Please note the grammar and its styling.  I believe we are within the confines of fair use, here.

Please note the additional page:  inactiveLink.html.  I'm suggesting that we use it for internal links where we do not yet have a good stub to which to point.  We could simply point to the applicable chapter, but then we'd later have to carefully review every link to determine whether  it actually is meant to point to the chapter as a whole or whether it is meant to point to a particular location within the chapter.  To simplify that work in the future, I'm instead pointing at the inactive link page.

mattrips committed:
Adding Content to Statements Chapter...

... added styling, auto numbering, etc.

mattrips committed:
Make Nav Bar Scrollable

When the choices in the nav bar overflowed the viewport height, they were not scrollable.  Added styling to header tag to fix this issue.

dabrahams commented on assets/css/grammar.css:
```diff
@@ -1,49 +1,16 @@
-    white-space: nowrap
-}
-
-
```
Why all this deletion?

mattrips replied:
When I created grammar.css (which is now too narrowly named), I created some classes, etc., that I never actually used.   I removed them, now, to eliminate clutter and unused code.

dabrahams replied:
Ah, I see.  I wrongly thought this was the main .css file.

dabrahams commented on assets/css/grammar.css:
```diff
@@ -86,12 +48,15 @@ div.Swift {
 .grammar .syntax-group {
     margin-bottom: 10px;
 }
+
```
Anything you wanna do in a new class namespace (`grammar`) is fine with me.

dabrahams commented on assets/css/grammar.css:
```diff
@@ -86,12 +48,15 @@ div.Swift {
+
+/* admonition */
+
 .admonition {
```
An admonition is a general, non-grammar-specific thing, so shouldn't be in grammar.css it seems to me.

It would be better if each logical change to the general formatting were a separate commit, with a commit message that explained what it was achieving and why.  It's hard to evaluate the effects of CSS changes from looking at diffs, even for CSS experts (which is not me) because of the way different declarations interact.

dabrahams commented on assets/css/grammar.css:
```diff
@@ -103,6 +68,9 @@ div.Swift {
 
+
+/* code */
+
```
code is also not a grammar-specific thing, so it seems to me this belongs in a different .css file.

dabrahams commented on assets/css/grammar.css:
```diff
@@ -129,33 +97,69 @@ body pre,body code{
 
 p>code,li>code,dd>code,blockquote>code,td>code{
     color:#000;
```
That declaration is pretty esoteric; it would be good to have a textual comment here saying what it applies to (I looked it up, that's a "parent" selector) and why you are explicitly styling only the code that has that particular set of parents.

dabrahams commented on assets/css/grammar.css:
```diff
@@ -129,33 +97,69 @@ body pre,body code{
 h2 {
-    counter-reset: section 0;      /* Set section to 0 */
+    text-indent: -40px;
+    counter-reset: section;
```
Auto-section numbering seems like a great idea, though of course there will be the temptation for people to use the numbers as though they are stable even when everything is in flux.  I don't know if there's a good mitigation strategy.  Also, we should number chapters as well, and sections should be nested inside chapters.  Any idea how to do that?

As for Indenting and shrinking of headers, I'm “agin' it,” except for chapter titles (H1) which should be bigger.  I think indenting is going to limit the horizontal space we have available and shrinking will make the document seem “over-styled”. The “nav bar” gives an overview of the document structure.  The working draft PDF formatting is pretty good, IMO.  http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/n4861.pdf

dabrahams replied:
How about auto-numbering paragraphs too, a la http://eel.is/c++draft ?  Oh, I see that you've done something like that, somehow!

The main advantage here would be (assuming we can achieve it) every paragraph would be individually linkable-to.  I don't see any anchors being generated by your stuff, and I'd be surprised if CSS can add them—but what do I know?.  Of course that wouldn't necessarily be a stable link especially during  early drafting, but it's hugely useful.

Speaking of which, we need to figure out how to make every [sub...]section heading a link target.  There's a way; I just don't know what it is.

dabrahams commented on assets/css/grammar.css:
```diff
@@ -129,33 +97,69 @@ body pre,body code{
+
 
+/* item numbering */
+.item {
```
Huh, what do you use "item" for?

dabrahams commented on assets/css/grammar.css:
```diff
@@ -129,33 +97,69 @@ body pre,body code{
 
-
-
 /* section numbering */
```
More stuff that doesn't seem to belong in grammar.css

dabrahams commented on chapters/0900 Statements.md:
```diff
@@ -9,6 +9,67 @@ is intended to describe aspects of Swift analogous to what's found in
 {% endcomment %}
 
+
+<div class="admonition grammar">
```
I guess you used raw HTML here because you could just copy it from the Swift website?  That totally makes sense.

Otherwise, I mean, there's https://kramdown.gettalong.org/quickref.html#html-elements
https://digitaldrummerj.me/styling-jekyll-markdown/ and https://idratherbewriting.com/learnapidoc/pubapis_markdown.html#markdown-and-static-site-generators that can be used to do some of this in a markdown context.  Also you could consider using a liquid variable for repeated stuff like `<span class="arrow"> → </span> `

It occurs to me we should also settle on an indentation standard 😝

dabrahams commented on chapters/0900 Statements.md:
```diff
@@ -17,31 +78,317 @@ is intended to describe aspects of Swift analogous to what's found in
 </div>
 
+<p class="item"> </p>
+*See*, <a href="0800 Expressions">Expressions</a>.
```
```suggestion
*See* <a href="0800 Expressions">Expressions</a>.
```

dabrahams commented on chapters/0900 Statements.md:
```diff
@@ -17,31 +78,317 @@ is intended to describe aspects of Swift analogous to what's found in
+</div>
+
+<p class="item"> </p>
+*See*, <a href="1000 Declarations">Declarations</a>. 
```
```suggestion
*See* <a href="1000 Declarations">Declarations</a>. 
```

dabrahams commented on chapters/0900 Statements.md:
```diff
@@ -17,31 +78,317 @@ is intended to describe aspects of Swift analogous to what's found in
+	</div>
+</div>
+
+<p class="item"> </p>
```
It would be good to closely follow the pattern of http://eel.is/c++draft/stmt.select in this area of the document.

dabrahams replied:
In fact, when filling in the [sub]sections of each chapter, IMO it would be a good idea to start by using the C++ draft for guidance, and include a link to each corresponding part.   There's so much of value there in terms of just *how to talk about things* in a rigorous way…

mattrips commented:
I snuck it in without a PR. My bad. 

Sent from my iPhone

> On Jul 19, 2020, at 1:40 PM, Dave Abrahams <notifications@github.com> wrote:
> 
> ﻿
> @dabrahams commented on this pull request.
> 
> In assets/css/grammar.css:
> 
> > -    padding: 4px 0 2px 10px;
> -    margin: 10px 0 21px;
> -    line-height: 22px;
> -}
> -
> -.highlight .gi { /* code placeholder capsule */
> -    background-color: rgba(233, 239, 250, 1);
> -    border: 1px solid rgba(196, 212, 241, 1);
> -    border-radius: 10px;
> -    color: rgba(0, 0, 0, 1);
> -    padding: 2px 6px;
> -    font-style: inherit;
> -    white-space: nowrap
> -}
> -
> -
> Ah, I see. I wrongly thought this was the main .css file.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

dabrahams commented on chapters/0900 Statements.md:
```diff
@@ -17,31 +78,317 @@ is intended to describe aspects of Swift analogous to what's found in
+#### Condition List
+
+A condition list is used in branch statements and while statements to determine 
+the path of execution.
```
The C++ standard uses the term "[flow of control](http://eel.is/c++draft/stmt.select#1)"

dabrahams commented on chapters/0900 Statements.md:
```diff
@@ -17,31 +78,317 @@ is intended to describe aspects of Swift analogous to what's found in
+<p class="item"> </p>
+A condition list consists of a list of one or more conditions. There are four 
+types of conditions:
+* an expression that evaluates to a `Bool`
```
Not quite the right way to talk about it.  See http://eel.is/c++draft/stmt.pre#6.  It's not that it *evaluates* to a `Bool` but that it *can be evaluated* as `Bool`.  But I'm not sure that's the way to say it.  The whole C++ notion of “[contextually converted](http://eel.is/c++draft/conv#def:conversion,contextual_to_bool)” is needed because of C++'s liberal implicit conversion, so we don't need anything quite that complicated.  But the basic idea is nearly the same: we want a corresponding “contextually *evaluated as* `Bool`,” like:

> Certain language constructs require that an expression be evaluated as a Boolean value. An expression *E* appearing in such a context is said to be *contextually evaluated as `Bool`* and is well-formed if and only if the expression `{ let `*t*`: Bool = `*E*`; return` *t* `}()` would be well-formed in its place, for some unique identifier *t* not otherwise used in the program.

Can't use `as` here unfortunately due to https://forums.swift.org/t/strange-asymmetry-bug-or-feature/38481

dabrahams replied:
Note: the ability to define well-formedness and semantics in terms of syntactic transformations like this is an extremely powerful simplifying tool and we should use it whenever possible.

dabrahams commented on chapters/0900 Statements.md:
```diff
@@ -17,31 +78,317 @@ is intended to describe aspects of Swift analogous to what's found in
+</div>
+
+<p class="item"> </p>
+A condition list consists of a list of one or more conditions. There are four 
```
I don't think we need to say this.  You're just restating what's in the grammar.  For example, http://eel.is/c++draft/stmt.block#nt:compound-statement doesn't say a compound statement is one or more statements.

dabrahams commented on chapters/0900 Statements.md:
```diff
@@ -17,31 +78,317 @@ is intended to describe aspects of Swift analogous to what's found in
+* a compiler `#availability` condition
+
+<p class="item"> </p>
+A condition list evaluates to `true` if all conditions of the condition list 
```
“Evaluates to” starts to imply a that the condition list (and indeed, each of its conditions) is an expression that could be used in any expression context, but it isn't.  Can we use "is satisfied?”

dabrahams commented on chapters/0900 Statements.md:
```diff
@@ -17,31 +78,317 @@ is intended to describe aspects of Swift analogous to what's found in
+evaluate to `true`.  
+
+<p class="item"> </p>
+Conditions of a condition list are evaluated in sequence, from left to right.  
```
```suggestion
Conditions of a condition list are evaluated in their syntactic order.  
```

dabrahams commented on chapters/0900 Statements.md:
```diff
@@ -17,31 +78,317 @@ is intended to describe aspects of Swift analogous to what's found in
+
+<p class="item"> </p>
+If an expression is used as a condition, the expression must evaluate to a 
+`Bool`.  Use of any other type of expression is an error.  [Note: The C `bool` 
```
Not objecting too hard, but the note seems a little out of place here.  

Heh, we need chapters for C and Objective C interop (just filed #8 and #9).

dabrahams commented on chapters/0900 Statements.md:
```diff
@@ -17,31 +78,317 @@ is intended to describe aspects of Swift analogous to what's found in
+##### Optional Binding Condition
+
+An optional binding condition attempts to initialize a variable from an instance 
+of `Optional`.  If the initialization succeeds, the variable is available within
```
I suspect you may be able to use some syntactic transformation here, though I haven't thought it through.

dabrahams commented on chapters/0900 Statements.md:
```diff
@@ -17,31 +78,317 @@ is intended to describe aspects of Swift analogous to what's found in
+</div>
+
+<p class="item"> </p>
+The `if` keyword is dependent upon the `Bool` type declared in the standard 
```
The meaning of this statement is very non-obvious to me.  Is it relevant where `Bool` is declared?  In what sense can a keyword (syntactic element) be “dependent upon” a type?

mattrips replied:
One of the unusual (IMO) things about Swift is the core language's dependency upon the standard library.  Things like `if` and `for`-`in` loops simply cannot work without certain types vended by the standard library.  Where that occurs, it seems worth noting.

dabrahams commented on chapters/0900 Statements.md:
```diff
@@ -17,31 +78,317 @@ is intended to describe aspects of Swift analogous to what's found in
+The `if` keyword is dependent upon the `Bool` type declared in the standard 
+library.
+<p class="item"> </p>
+The code block of an `if` statement is executed if the condition list of the 
```
```suggestion
The code block of an `if` statement is executed if and only if the condition list of the 
```

dabrahams replied:
I think we can model this section on http://eel.is/c++draft/stmt.if#1

dabrahams commented on inactiveLink.html:
```diff
@@ -0,0 +1,21 @@
+<head>
```
Not a fan of this file, at least in its current form.

If we have a bunch of future link targets we should establish standard link target names for them, and in the meantime the links should take you to an entry in a list of descriptions of what each link is supposed to point at.  I'm not sure what the best way to engineer our site to do that would be.

Can we not stick to markdown instead of raw HTML?

dabrahams reviewed (changes requested):
Once again, 1000 thanks for bravely diving in.  LMK if you have any questions about all this; I hope my many comments aren't discouraging.  I think this is a great start!

mattrips committed:
Merge branch 'master' into StatementsWorkingBranch

mattrips committed:
Adjusted positioning of opt notation in grammar.

The tightness of "opt" relative to the grammar element that it marks varies depending upon whether it is marking code, which is styled with padding, or the name of a grammar part, which is not styled with padding.  We could style the names of grammar parts to add padding, but it was easier to simply add a class (grammar sub.loosen) to override the normal behavior of (grammar sub).

mattrips committed:
Inserted Protocol Chapter and Renumbered .

Renumbering chapter files seems like a messy and bad practice.  But, in these early days, perhaps it isn't so bad.

mattrips committed:
Revert "Inserted Protocol Chapter and Renumbered ."

This reverts commit 8df9400ca5725601b0a67c7508882f42182447df.

mattrips committed:
Revert "Adjusted positioning of opt notation in grammar."

This reverts commit d6e6537e04aacbd6200db0ed9ea51d917f90bbe2.

mattrips committed:
Fix for tighness of opt annotation; added comments.

In some cases, the "opt" grammar annotation is too tight to item it modifies.  As a temporary fix, adding a class (grammar sub.loosen) that loosens the spacing.

Also, added numerous comments to explain the purpose of each style and to note which styles  are to be moved to new files.

mattrips committed:
Moved styles for headings into separate css file

mattrips committed:
Moved navigation scrollbar styling to styles.scss

mattrips committed:
Merge pull request #10 from dabrahams/CssModifications

Moving the CSS Modifications into their Own Branch and PR
